### PR TITLE
[MM-53019] Add file storage permission check to workspace health dashboard

### PR DESCRIPTION
**Pull request by hanzei**
#### Summary
The server already checks if the file storage is available on startup. To surface that information even more, this PR introduces a file storage check to the workspace health dashboard.

I've also filed a ticket to add file storage information to the Support Packet: https://mattermost.atlassian.net/browse/MM-54318

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-53019

#### Screenshots
Screenshot from 2023-08-29 11-26-05

#### Release Note
```release-note
Add file storage permission check to workspace health dashboard
```

### CODE REVIEW

**Comment by hanzei** [comment]
Blocked by https://community.mattermost.com/core/pl/jom7f6xtaidk8rigpnjccafw4o

File: webapp/platform/client/src/client4.ts
Comment (hanzei) [inline webapp/platform/client/src/client4.ts:2353]: `time` is an unknown field for the server, hence it gets removed.

File: webapp/platform/types/src/client4.ts
Comment (hanzei) [inline webapp/platform/types/src/client4.ts:26]: I picked this solutions based on the discussion in https://community.mattermost.com/core/pl/frhr5twtrpyt9d54ze4djero6c

File: webapp/channels/src/utils/constants.tsx
Comment (hanzei) [inline webapp/channels/src/utils/constants.tsx:1102]: TODO: This still needs to be set up.

File: server/channels/api4/system.go
Comment (hanzei) [inline server/channels/api4/system.go:156]: Given how expensive the check is it should be guarded properly.

File: webapp/platform/types/src/client4.ts
Comment (hmhealey) [inline webapp/platform/types/src/client4.ts:18]: I think it might be better if you just add a JSDoc comment to just the `ignoreStatus` field instead of adding it to `Options`

File: webapp/platform/client/src/client4.ts
Comment (hmhealey) [reply webapp/platform/client/src/client4.ts:2353]: Is there any caching on the ping API that this was being used to bust? That's what it looks like it was doing before

File: webapp/platform/client/src/client4.ts
Comment (hmhealey) [inline webapp/platform/client/src/client4.ts:2348]: Do we need to add the device ID here? I don't think that's used on web. 0/5 if it's just for consistency though

File: webapp/channels/src/components/admin_console/workspace-optimization/dashboard_checks/config.ts
Comment (hmhealey) [inline webapp/channels/src/components/admin_console/workspace-optimization/dashboard_checks/config.ts:85]: This is definitely not specific to this PR, but it feels really weird that the ping API is used for these sort of checks. I know we've added more information to it over time, but I think it was originally just to check if the server was alive so mobile/desktop would know it was a valid server, and I wouldn't expect it to be used for stuff like this

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: webapp/platform/client/src/client4.ts
Comment (hanzei) [reply webapp/platform/client/src/client4.ts:2353]: I can't find any caching on the server endpoint :shrug: I don't know why the  `time` field was added.

File: webapp/platform/types/src/client4.ts
Comment (hanzei) [reply webapp/platform/types/src/client4.ts:18]: I wasn't able to find a way to only add a comment to `ignoreStatus` that is valid JSDOC. Can you point me to documentation that explains the proper syntax?

File: webapp/platform/client/src/client4.ts
Comment (hanzei) [reply webapp/platform/client/src/client4.ts:2348]: 1/5 that I like the consistency with the server code.

File: webapp/channels/src/components/admin_console/workspace-optimization/dashboard_checks/config.ts
Comment (hanzei) [reply webapp/channels/src/components/admin_console/workspace-optimization/dashboard_checks/config.ts:85]: I would like to see the endpoint getting split into two:
1. One login-related one that also contains information like  `CanReceiveNotifications` or Android and iOS-supported versions.
2. A health check one, that is used for probing in Cloud and the dashboard.

**Review by pvev — APPROVED** [review]
Thank you @hanzei , your change LGTM and makes sense. Only one question left. :shipit:

**Comment by pvev** [comment]
@hanzei should this one also be updated? https://github.com/mattermost/mattermost/blob/8fdba52a66e6199b29b483a19828fae688bdf1ec/webapp/channels/src/components/admin_console/license_settings/license_settings.tsx#L205

**Comment by hanzei** [comment]
What exactly needs an update in the license settings? The code there simply ping the server to check if it is up again.

**Comment by pvev** [comment]
> What exactly needs an update in the license settings? The code there simply ping the server to check if it is up again.
I think TS will complain cause you are not passing a value to the first parameter which is not optional in your definition.

File: webapp/platform/types/src/client4.ts
Comment (hmhealey) [reply webapp/platform/types/src/client4.ts:18]: It's just the same `/** */` comment style as this, but you add it to just a single field on the object. We've mostly done that for component props, but there's some places that we use it in this package like [this one](https://github.com/mattermost/mattermost/blob/master/webapp/platform/types/src/teams.ts#L57)

File: webapp/platform/client/src/client4.ts
Comment (hmhealey) [reply webapp/platform/client/src/client4.ts:2348]: That makes sense to me. I forgot this was also in the client package which we want other people to use

File: webapp/channels/src/components/admin_console/workspace-optimization/dashboard_checks/config.ts
Comment (hmhealey) [reply webapp/channels/src/components/admin_console/workspace-optimization/dashboard_checks/config.ts:85]: Yeah, that's what I'd expect too. That's obviously not something we can change now though

File: webapp/platform/client/src/client4.ts
Comment (hmhealey) [reply webapp/platform/client/src/client4.ts:2353]: I guess that's probably a legacy thing. We used to be a lot more aggressive with adding `Cache-Control` headers to things

**Review by hmhealey — APPROVED** [review]
I replied to the JSDoc comment from above, but that's not blocking

File: webapp/platform/types/src/client4.ts
Comment (hanzei) [reply webapp/platform/types/src/client4.ts]: Thanks for the hint. But I still can't get VSCode to show the comment.


Screenshot from 2023-10-04 09-53-29


Screenshot from 2023-10-04 09-53-23

It's surely unrelated to the PR, but do you know if any specific extension in needed?

**Comment by hanzei** [comment]
@lindy65 I've left QA test steps in the JIRA ticket. Please let me know if anything is unclear.

**Comment by DHaussermann** [comment]
@hanzei I'm not sure I can test this conclusively. Even without disabling the permissions to the data folder the dashboard shows and issue. 
On a fresh deploy (after `make-nuke`) with this branch, I see an issue with the file storage. 
 
image

- File storage is set to use local file system and I see nothing obviously wrong (it's pointed to `./data/`)
- For what it's worth the permissions on the data folder are still normal at `drwxr-x` when I run `stat data`  they have not changed yet. 

- Strangely server logs show no `Error` occur aside from no license applied
- No mention of storage I can see in the logs either since the start-up

1. @hanzei Is the issue here that it's telling me something is wrong when there is nothing to adjust or am I missing a legit issue? Can you suggest how I can narrow down the issue?

2. The `Learn more` button points here https://mattermost.com/pl/configure-file-storage which shows a Page not found in the MM help. Maybe @cwarnermm would know if there is a valid place to direct people for help with file storage or if that page should just work.    
image

**Comment by cwarnermm** [comment]
@emdecr - The permalink ``https://mattermost.com/pl/configure-file-storage`` appears to be missing from the [master Marketing spreadsheet](https://docs.google.com/spreadsheets/d/1O601H_A0IM8pR3FOfue29_C8flGV7xK3ts-tJUVDHHg/edit#gid=0). The resulting page isn't found via the docs site when accessed in-product. 

Are you open to creating this permalink and setting it to take users to ``https://docs.mattermost.com/configure/environment-configuration-settings.html`` please?

**Comment by emdecr** [comment]
@cwarnermm Thanks for the flag! Created the /pl/ and added the item to the master sheet 👍

**Comment by cwarnermm** [comment]
@DHaussermann - With the permalink being created/live, the in-product link should now work. Can you re-test please?

**Comment by hanzei** [comment]
@DHaussermann Good catch, the code was indeed faulty. I've fixed the bug in 234bd2b5c3 and the check now works correctly. PTAL.

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- When the data folder (or which ever folder is set as the Local Storage Directory) is working as expected, there is no longer a configuration problem shown. 
- When write access is removed Workspace Optimization shows a configuration problem on webapp refresh 
- When write access is restored, Workspace Optimization shows no more configuration problem on webapp refresh  
- note: I also checked the behavior if you point the Local Storage Directory to a directory that doesn't exist but this case was not applicable as the directory will get created so there is no configuration problem to solve.

LGTM!

**Pull request merged**
